Separated out logging to per deployment logging.

saisei opened the pull request:
Fusor logging can be accessed at ::Fusor.log
Created init_log.rb so logging are loaded in engine's initializer. 
Added base fusor action classes that would initialize deployment specific logging.
Modified existing action classes so it inherits from base Fusor classes.

Note: This is identical to pull request #516, except it's cleaner and does not contain any merge conflict mess.

jwmatthews commented:
@saisei there are a few rubocop warnings we'll need to clean up, see the link to the travis run.  Additionally you'll need to come up with a solution for the logging to run with unit tests, as travis is complaining of "/var/log/foreman/foreman.log" doesn't exist.

If you want to modify the travis env to create the /var/log/foreman, you can look at/modify the below file:
https://github.com/fusor/fusor/blob/master/.travis.sh

Or you could add some code to the logger where if the expected file doesn't exist it will create it.

saisei commented:
Updated the commits to include changes that fixes rubocop errors and offenses. 

Also added a card in trello to keep track of future improvements for logging capabilities: https://trello.com/c/Cg7UjheC/695-improve-existing-logging-capabilities-in-fusor-server

saisei commented:
@jwmatthews please test out the code to determine if this can be merged.

daviddavis commented on server/lib/init_log.rb:
```diff
@@ -0,0 +1,24 @@
+
+    #@log.detach_all
+
+    if Rails.env.production?
```
I would advise against this. This can cause errors to go undetected in development/test. If you're trying to get this to work in Travis/test environment, it'd be better to mock or stub this method. Or even better, use [FakeFS](https://github.com/defunkt/fakefs) which foreman provides.

jmrodri replied:
+1 use Rails.root, it does work (at least from what I've seen) in Travis.

daviddavis commented on server/lib/init_log.rb:
```diff
@@ -0,0 +1,24 @@
+        @log.attach(@default_log_file)
+      else
+        Dir.mkdir("/var/log/foreman/#{deployment.name}-#{deployment.id}") unless File.exist?("/var/log/foreman/#{deployment.name}-#{deployment.id}")
+        @log.attach("/var/log/foreman/#{deployment.name}-#{deployment.id}/deployment.log")
```
I wouldn't hardcode this to `/var/log/foreman`. If you use `File.join(Rails.root, 'log')`, this will work in all environments.

Also, I raised [an issue about using the deployment.name in the filename in a previous PR](https://github.com/fusor/fusor/pull/485#discussion_r45390172). Not sure if you saw that.

cfchase replied:
for the deployment.name issues seems like we can do 1 of 2 things pretty easily.
1) just drop name and use id which should always be unique
or 2) strip out any potentially offensive chars.
```ruby
deployment.name.gsub!(/[^0-9A-Za-z.\-]/, '_')
```
